lib/Component.php

```diff
@@ -16,9 +16,14 @@ public function __construct($factory, $component_data) {
     $this->spec     = $component_data['spec'];
     $this->name     = $this->spec->name;
     $this->version  = $this->spec->version;
-    $this->location = $component_data['location'];
-    $this->subpath  = $component_data['subpath'];
-    $this->type     = isset($this->spec->type) ? $this->spec->type : 'backend';
+
+    if (array_key_exists('base_dir', $component_data)) {
```

Comment:
I think the important thing is that the developer knows what (s)he's doing and not accessing anything you don't know is there :)

I was on the other hand thinking about instead of base_dir having a container of data that only the factory is interrested in.  base_dir for example is only used in the factory-call getAsset.